### DISCUSSION TITLE
How to use async functions inside lambda?

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1424] [Q&A] [answered]
### Question

Hi All,

I am new to the async functions and trying to learn how to use them inside my NiceGUI app slowly.
I am trying to learn how to call an async function from a button click (like those lambda functions)?

The use case is, let's say the user fills some inputs and clicks a button, which then runs a database query and pulls some results. Now let's say I am converting the database query to async and then trying to call it from the button click. Below is a rough example of what I am trying to do. Please ignore the actual database query example, because that one is not needed, just to explain the flow.
```python
@ui.page('/')
async def index():
    username = ui.input('Username')
    ui.button('Submit', on_click=lambda: input_check())

    async def input_check():
        user = await dup_username(username.value)
        if user:
            ui.notify('This username is already taken', type='negative')
        else:
            ui.notify('This username is available', type='positive')


    async def dup_username(username):
        var = 'username'
        query = f'SELECT {var} FROM users WHERE {var} COLLATE NOCASE = ?'
        async with aiosqlite.connect(db_path) as conn:
            async with conn.execute(query, (username,)) as cursor:
                existing_username = await cursor.fetchone()
        return existing_username is not None
```

So the question is, in this case, the lambda function is just treating the  ```input_check``` function as a normal function right? Which means even though I make the ```input_check``` function async, the ```dup_username``` function is not really awaited, right?

So how should I correctly approach this? How to call these kind of async database queries using lambda functions of button clicks?  

Thanks a lot for all the help and guidance.

Regards,
Anindya

**Comment by falkoschindler** [comment, marked as answer]
Hi @Anindya088!

Your code should simply work. Even though the lambda returns an awaitable, NiceGUI's event handler automatically awaits the result. So the following two buttons both run and await `my_async_function`:

```py
async def my_async_function():
    await asyncio.sleep(1)
    ui.notify('Done')

ui.button('Run', on_click=my_async_function)
ui.button('Run', on_click=lambda: my_async_function())
```

**Reply by Anindya088** [reply]
Hi @falkoschindler ,

Oh great, thanks so much for the answer. I was not sure if calling just the function directly inside lambda will actually make the inside async function await or not. I was just reading in general on async functions etc. on Google, and what I understood was to run an async function from inside a normal (sync) function, we need to use ```asyncio.run``` or other event loop methods, and I did not understand how to work them out 😄 

But thanks for confirming that just the normal usage of lambda function will be sufficient in this case. I should be able to work out the rest now.

Thanks a lot,
Anindya
